partner_firstname/models.py

```diff
@@ -80,13 +83,20 @@ def _inverse_name(self):
         # Company name goes to the lastname
         if self.is_company or self.name is False:
             parts = [self.name, False]
-
         # Guess name splitting
         else:
-            parts = self.name.split(" ", 1)
-            while len(parts) < 2:
-                parts.append(False)
-
+            parts = self.name.split(" ")
+            if len(parts) > 1:
+                if self.company_id.names_order == 'first_last':
+                    first = parts[0]
+                    last = u" ".join(parts[1:])
+                else:
+                    first = parts[-1]
+                    last = u" ".join(parts[:-1])
+                parts = [last, first]
+            else:
+                while len(parts) < 2:
+                    parts.append(False)
         self.lastname, self.firstname = parts
```

Comment:
You should use `self.update()` here, since there's a possibility that constraint fails (if firstname="something" and lastname=False).